src/rust/engine/src/session.rs

```diff
@@ -332,3 +300,59 @@ impl Session {
     }
   }
 }
+
+///
+/// A collection of all live Sessions.
+///
+/// The `Sessions` struct maintains a task monitoring SIGINT, and cancels all current Sessions each time
+/// it arrives.
+///
+pub struct Sessions {
+  /// Live sessions. Completed Sessions (i.e., those for which the Weak reference is dead) are
+  /// removed from this collection on a best effort when new Sessions are created.
+  sessions: Arc<Mutex<Vec<Weak<InnerSession>>>>,
+  /// Handle to kill the signal monitoring task when this object is killed.
+  signal_task_abort_handle: AbortHandle,
+}
+
+impl Sessions {
+  pub fn new(executor: &Executor) -> Result<Sessions, String> {
+    let sessions: Arc<Mutex<Vec<Weak<InnerSession>>>> = Arc::default();
+    let signal_task_abort_handle = {
+      let mut signal_stream = signal(SignalKind::interrupt())
```

Comment:
LGTM. My only worry in the small was what happens when new gets called more than once in the same process (I don't think that can happen though from an in the large view of the code). Even if this occurred, the tokio docs says its ok and events just get delivered to all signal streams and thats fine if it happens since AsyncLatch will only fire once and not error after.